Gnupg2.2

Ionic opened the pull request:
Original description by @roederja:
>###### Description
>
>gnupg -> modern gnupg 2.2.1
>gnupg1 -> legacy gnupg 1.4.22 (used to be gnupg)
>gnupg2 -> obsolete, use gnupg instead
>gnupg21 -> obsolete, use gnupg instead

Ionic commented:
Rebased `.DS_Store` removals into previous commit.

l2dy commented on mail/gnupg/Portfile:
```diff
@@ -2,64 +2,77 @@
 
-name                gnupg
-version             1.4.22
+name                gnupg2
```
gnupg2 in gnupg Portfile?

l2dy commented on mail/gnupg2/Portfile:
```diff
@@ -2,14 +2,16 @@
 name                gnupg2
 version             2.0.30
-revision            2
+revision            1
```
Decreased revision?

Ionic commented:
Complete refactoring with (hopefully) easy-to-review separate commits.

Note that these changes will be squashed into one commit before actually merging this PR.

`gnupg1`: copy of old `gnupg` port.
`gnupg-stable`: copy + update to 2.2.1 of old `gnupg21` port.
`gnupg2`: obsolete, replaced by `gnupg-stable`.
`gnupg21`: obsolete, replaced by `gnupg-stable`.
`gnupg`: obsolete, if an old `gnupg` port is currently active upgrades to `gnupg1`. Otherwise denies installation and asks the user to install either `gnupg1` or `gnupg-stable`, recommending the latter.
`gpg-agent`: obsolete, replaced by `gnupg-stable` (ships `gpg-agent` internally with autospawn feature.)
`dirmngr`: obsolete, replaced by `dirmngr` (ships `dirmngr` internally with autospawn feature.)

Still todo: update all dependents.

Ionic commented:
Also pinging @lbschenkel, since he was active on the mailing list.

Ionic commented:
Dependencies updated, but **not** tested. Probably also needs more fine-tuning.

Ionic commented:
`gnupg-stable` now also installs a `gpg2` symlink in `{prefix}/bin/`.

Needed for some ports that explicitly want modern `GnuPG`.

Ionic commented:
And dependencies updated to hopefully reflect what the ports actually need. It's now more flexible and users can finally opt to use modern GnuPG versions even with keybase et. al.

Ionic commented:
Regarding the name: I really like `gnupg-stable`. If you all don't, I'll rename it to `gnupg2` (and get rid of that stub-port.)

`gnupg1` is probably a good idea, since it's canonically also used in other package systems, so I changed my mind about naming that `gnupg-legacy`.

Going forward, like in a year, I want to make `gnupg1` and `gnupg-stable` co-installable, with `gnupg1` only providing `1`-postfixed binaries. All dependencies upon `gnupg1` should then be dropped completely, so that only users that actually need 1.4 also install it. (Werner Koch, the project head, incidentally needed to install 1.4 for decrypting some old mails recently, that were not decryptable with modern GnuPG. So keeping 1.4 around as a port that nothing depends on for special tasks makes sense.)

lbschenkel commented:
Great work! I didn't check all commits, but it looks good. A few suggestions:
- Were I doing this, I would have attempted to do the `1`-postfix now. This will require more work of course, but I would have configured/patched the ports depending on `gnupg1` to invoke `gpg1`. (This will be consistent with other distros do, like Debian.) I would also have considered adding a non-default variant that installs a symlink to `gpg`, so in the rare case people are using just `gpg1`, don't want to upgrade, and want to invoke it the way they used to they can manually enable this variant. A description could be added to explain this.
- I really grew fond of the `gnupg1`/`gnupg`/`gnupg-devel` naming that was suggested on the list, which also seems to be the most popular choice. The advantages that I see are that they reduce the number of stub ports, remain faithful to the existing MacPorts naming guidelines and is also consistent with other distros. (But no matter the naming used, the changes being made here are a step forward and saner than what we currently have.)

lbschenkel commented on devel/gpgme/Portfile:
```diff
@@ -29,34 +29,15 @@ depends_build       port:pkgconfig
+                    patch-tests-gpg-Makefile.am.diff
 
 post-patch {
+    reinplace "s|@GPG@|${prefix}/bin/gpg2|" ${worksrcpath}/src/gpgme-config.in
```
I have made a separate comment, but I think it would be more elegant if `bin/gpg` is always the stable version and the few ports that depend on legacy invoke it via `bin/gpg1`. It looks backwards to me to have the ones depending on the "mainstream" one having to invoke a non-default binary.

Ionic replied:
For now, I want to keep it installing `gpg`. In about a year, I want to make it only install `gpg1` and have no other port depend on it, with it being co-installable with the other ports. That way, users that still need `gnupg1` can install it, but it's not pulled in by default by any other port. Essentially deprecation in a year.

For now, I'd like to keep them conflicting.

lbschenkel replied:
The part that I don't quite follow is what is gained by doing that one year from now compared to now. Less work? Potential less breakage? What are the ports that would pull gpg1 using the default set of variants?

I'm sure you have a good reason, though. I'm just curious to find out what I'm missing.

Ionic replied:
Merely to let people keep their setup (mostly, sans the agent and dirmngr support...) for a little bit longer if they are determined to. Those that chimed in with a `+1` already use the modern version, but those that want to stay on `gnupg1` would definitely give a `-1`.

Since we have to keep at least some stub ports around anyway, I find that a good balance between forcing users to the modern version and giving them a grace period.

I probably should add a notes message about that to the new `gnupg1` port, that I think of it, though.

Ionic replied:
Note added.

roederja replied:
I couldn't find any ports that depend on gnupg, but maybe my text search isn't working right.

roederja replied:
Regarding making gnupg1 and gnupg2 / -stable co-installable: Won't you run into the problem that the keystore formats are not compatible?

Re port naming: I would keep gnupg2 as what you now named gnupg-stable, then you can in a year from now move that into gnupg and make gnupg2 obsolete. If necessary we can add a gnupg-devel port to be consistent with macports conventions, but we won't have a -stable port that we already know we have to deprecate in a year.

Ionic replied:
Yes, keystore formats differ, but that's fine.

The co-installable `gnupg1` port will be provided as a courtesy, but not referenced within MacPorts any longer. Thus, it's only useful to decrypt old data, which is not possible with the modern GnuPG version (cases like these exist) and we assume that users either have an old keyring storage format backed up, or that they know how to export and import public and private keys (which can be used to exchange keys between the two formats.)

Yeah, calling it `gnupg2` would remove a stub-port now, but stub-ports are cheap. We don't maintain these, they just sit there for a year and get removed.

Anyway, I can rename it to `gnupg2`.

Ionic commented:
>I really grew fond of the gnupg1/gnupg/gnupg-devel naming that was suggested on the list, which also seems to be the most popular choice. The advantages that I see are that they reduce the number of stub ports, remain faithful to the existing MacPorts naming guidelines and is also consistent with other distros. (But no matter the naming used, the changes being made here are a step forward and saner than what we currently have.)

Nope, can't do `gnupg1`/`gnupg`/`gnupg-devel`. If anything, I can do `gnupg1`/`gnupg2`/`gnupg-devel`. The reason is that `gnupg` has to stay around as a sub-port, that allows users to either stay on `gnupg1` for now (if `gnupg` is active when they attempt to upgrade it) and otherwise tells users to that they have a choice between `gnupg1` and `gnupg-stable` (or `gnupg2`, if users prefer that name), with the latter strongly recommended.

I don't like `gnupg2` too much, because of it containing a version number. If upstream decided to switch the numbering scheme and release version 3 in two months, we'll have another stub port. Hence me opting for `gnupg-stable`.

This said, I **can** replace the `gnupg` stub-port with the mainline version in a year (we generally assume that users upgrade their installation at least once a year. Keeping stub-ports around for at least a year provides a clean upgrade path in this case.)

roederja commented on mail/gnupg-stable/Portfile:
```diff
@@ -0,0 +1,83 @@
+
+PortSystem          1.0
+
+set my_name         gnupg
```
What's the deal with my_name?

Ionic replied:
That's a common pattern used when the port name does not match the software name (for instance when capitalization differs.) `git grep my_name` will provide you with a lot of examples.

Normally, when you change the `name` variable, you also implicitly change other variables (think of `dist*` variables for instance.)

If the port name does not match the software name, you have to override settings that normally are inferred via `${name}`, thus losing the mentioned benefit.

By employing a helper variable like `my_name`, this behavior can be retained.

roederja commented on security/openpgpsdk/Portfile:
```diff
@@ -21,7 +21,7 @@ checksums           md5     a2fe87ee864e6bcc0d5ea3db20e0f57a \
                     rmd160  9260e464f8f9b385d0cc8348d20e1c264856cee3
 
-depends_build       bin:perl:perl5 bin:gpg:gnupg
+depends_build       bin:perl:perl5 path:bin/gpg:gnupg-stable
```
Why do you use path: dependencies. This basically disallows installing your own version of gnupg outside of macports.

Ionic replied:
No, users can still install GnuPG outside of MacPorts - MacPorts will just not use it. And that's exactly the point. MacPorts strives to provide a self-contained system, save for the base system shipped by Apple. Since Apple does not ship GnuPG, these dependencies should never have been `bin:`-style dependencies in the first place. We do want to pull in our own version, over which we have control. Stuff that users install is outside of our control. We don't know what version that might be, we have no idea how it might have been patched, we don't even know if that's **really** GnuPG.

Schamschula commented:
I'm seeing a bit of an issue with the new naming scheme, even though it makes sense, as it is not in common use under MacPorts.

The closest thing, in terms of naming, I see in MacPorts is octave, there we have:
- octave-legacy (3.8.x - obsolete)
- octave (4.2.1) current
- octave-devel (tip - 4.3.0)
- octave-devel-rc (release candidate. Currently pointing to 4.2.1rc)
- octave-devel-release (same as octave, but built from hg)

 I'm currently similarly rejiggering the apache ports:
- apache (1.3.x - marked as obsolete)
- apache2 (currently 2.2.x, soon to be 2.4.x)
- apache20 (2.0.x - soon to be marked as obsolete)
- apache22 (soon to be 2.2.x)
- apache24-devel (2.4.x, soon to be apache2)
Note: I'm staying with the usual numbering scheme.

So for gnupg I'm looking at:
- gnupg1 (no problem here)
- gnupg2 (should be current, 2.2.1)
- gnupg21 (obsolete)

However, I think users will soon come to terms with any of these naming schemes.

Ionic commented:
Updated against current master branch and renamed `gnupg-stable` to `gnupg2`.

l2dy commented on mail/lbdb/Portfile:
```diff
@@ -51,7 +52,7 @@ post-destroot {
+    depends_lib-append      port:gnupg1
     configure.args-replace  --without-gpg --with-gpg
 }
 
```
Down below in gpg2 variant, `--with-gpg=${prefix}/bin/gpg2` should be changed to `--with-gpg=${prefix}/bin/gpg` like other ports.

P.S. GitHub doesn't allow me to directly comment there.

Ionic replied:
>Down below in gpg2 variant, --with-gpg=${prefix}/bin/gpg2 should be changed to --with-gpg=${prefix}/bin/gpg like other ports.

Nope, because I want the `gpg2` variant to explicitly depend upon a modern GnuPG port, not to be also satisfied by `gnupg1`. This, for instance, is also true for the `gpgme` port (which is a useful wrapper for modern GnuPG.)
I could hardcode `gnupg2` here, but I'd like to keep it a `path:`-based dependency in case we want to add a `gnupg-devel` port in the future for testing versions.

> P.S. GitHub doesn't allow me to directly comment there.

You can only add comments in or around changed hunks, unfortunately. I usually also do what you did - add a comment as close to the location I intended to when I hit this problem.

l2dy replied:
`depends_lib` should remain as is so it explicitly depends upon a modern GnuPG port, I'm just saying that the symlink is only there for compatibility so we should avoid calling it and change `configure.args`. `gnupg2` would conflict with any port that contains `${prefix}/bin/gpg` so this shouldn't be a problem.

Ionic replied:
Ah, you mean the configure argument. Yeah, that was how I implemented it initially, but I don't plan to drop the `gpg2` symlink any time soon and I'd rather see a build failure and a runtime failure if users decide to forcefully deactivate the modern GnuPG version and install the legacy one than it tagging along - potentially in a faulty way.

After thinking that through, I reverted back to the current implementation which feels safer to me.

Ionic commented:
Okay, since most people are happy with the proposed changes: what do we do about the dependents? I have changed them accordingly to the new layout, but didn't actually test installing and running all these ports that depend upon some `gnupg` port. Common courtesy would say to do that... or alternatively to hope that everything works and have port maintainers handle problems if any should come up.

lbschenkel commented:
I am the maintainer for `gpgme` and I'm OK with the changes. In case there any regressions pop up I'll handle those.

Ionic commented:
Yeah, but not all dependents. :)

I'll install all packages with either of the new packages and check that building them works. No runtime tests.

This list will be updated.

 - [ ] `gpgme`
   - [ ] w/ `gnupg2` 
 - [ ] `lbdb`
    - [ ] w/ `gnupg1` (`+gpg`)
    - [ ] w/ `gnupg2` (`+gpg2`)
  - [X] `signing-party`: only run-time dependency
  - [ ] `fwknop-client`
    - [ ] w/ `gnupg1`: doesn't work, dependency on `gpgme`
    - [ ] w/ `gnupg2`
  - [X] `tin-devel` (`+gpg`): only run-time dependency
  - [X] `tin` (`+gpg`): only run-time dependency
  - [ ] `p5-gnupg-interface` (only `p5.26-gnupg-interface`)
    - [ ] w/ `gnupg1`
    - [ ] w/ `gnupg2`
  - [ ] `p5-mail-spamassassin` (only `p5.26-mail-spamassassin`)
    - [ ] w/ `gnupg1`
    - [ ] w/ `gnupg2`
  - [ ] `p5-module-signature` (only `p5.26-module-signature`)
    - [ ] w/ `gnupg1` (`+gnupg1`)
    - [ ] w/ `gnupg2` (`-gnupg1`)
  - [X] `py-gnupg`: only run-time dependency
  - [X] `keybase`: only run-time dependency
  - [ ] `openpgpsdk`
    - [ ] w/ `gnupg1`
    - [ ] w/ `gnupg2`
  - [X] `pass`: only run-time dependency
  - [ ] `pwman`
    - [ ] w/ `gnupg1`
    - [ ] w/ `gnupg2`
  - [ ] `duplicity`
    - [ ] w/ `gnupg1`
    - [ ] w/ `gnupg2`
  - [ ] `zeroinstall-injector`
    - [ ] w/ `gnupg1`
    - [ ] w/ `gnupg2`

Ionic commented:
Everything looks good according to my tests.

I'll rebase this branch against current master, including squashing, and will merge it into master some time tonight.
